core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -506,13 +506,59 @@ private[spark] class TaskSetManager(
    * Get the level we can launch tasks according to delay scheduling, based on current wait time.
    */
   private def getAllowedLocalityLevel(curTime: Long): TaskLocality.TaskLocality = {
-    while (curTime - lastLaunchTime >= localityWaits(currentLocalityIndex) &&
-        currentLocalityIndex < myLocalityLevels.length - 1)
-    {
-      // Jump to the next locality level, and remove our waiting time for the current one since
-      // we don't want to count it again on the next one
-      lastLaunchTime += localityWaits(currentLocalityIndex)
-      currentLocalityIndex += 1
+    // Remove the scheduled or finished tasks lazily
+    def hasNotScheduledTasks(taskIndexes: ArrayBuffer[Int]): Boolean = {
+      var indexOffset = taskIndexes.size
+      while (indexOffset > 0) {
+        indexOffset -= 1
+        val index = taskIndexes(indexOffset)
+        if (copiesRunning(index) == 0 && !successful(index)) {
+          return true
+        } else {
+          taskIndexes.remove(indexOffset)
+        }
+      }
+      false
+    }
+    // It removes the empty lists after check
+    def hasMoreTasks(pendingTasks: HashMap[String, ArrayBuffer[Int]]): Boolean = {
+      val emptyKeys = new ArrayBuffer[String]
+      val hasTasks = pendingTasks.exists{
+        case (id: String, tasks: ArrayBuffer[Int]) =>
+          if (hasNotScheduledTasks(tasks)) {
+            true
+          } else {
+            emptyKeys += id
+            false
+          }
+      }
+      emptyKeys.foreach(x => pendingTasks.remove(x))
```

Comment:
Ah yes, I see now.  So, leave it as `id` and slip in the `key could be executorId, host, or rackId` comment.